Add hostname options to Keycloak CR

mabartos opened the pull request:
Closes #14395

mabartos commented on operator/src/main/java/org/keycloak/operator/controllers/KeycloakDeploymentConfig.java:
```diff
@@ -165,7 +153,11 @@ public void configureTLS() {
+
+        var kcPort = keycloakCR.getSpec().getHostnameSpec().getPort();
+        if (kcPort == null) {
+            kcPort = keycloakCR.getSpec().isHttp() ? Constants.KEYCLOAK_HTTP_PORT : Constants.KEYCLOAK_HTTPS_PORT;
```
@vmuzikar In the description of the associated issue is mentioned that the `INSECURE-DISABLE` functionality should be removed. Does it mean that we're gonna support only the production-ready environment with Operator? Will there be no support from the developer's perspective?

vmuzikar replied:
jFTR – this was discussed offline and will be reflected in docs.

vmuzikar commented on operator/src/main/java/org/keycloak/operator/crds/v2alpha1/deployment/KeycloakSpec.java:
```diff
@@ -67,18 +62,10 @@ public class KeycloakSpec {
-    public boolean isHostnameDisabled() {
-        return this.hostname.equals(Constants.INSECURE_DISABLE);
-    }
+    @NotNull
```
I think, in theory (a future dev mode or something) this can actually be `null`, right?

vmuzikar commented on operator/src/main/java/org/keycloak/operator/crds/v2alpha1/deployment/spec/HostnameSpec.java:
```diff
@@ -0,0 +1,106 @@
+@Buildable(editableEnabled = false, builderPackage = "io.fabric8.kubernetes.api.builder")
+public class HostnameSpec implements Serializable {
+
+    @NotNull
```
The `hostname` can now be `null` and is therefore optional (in case someone e.g. sets `hostname-strict=false`).

vmuzikar commented on operator/src/main/java/org/keycloak/operator/controllers/KeycloakIngress.java:
```diff
@@ -79,6 +83,7 @@ private Ingress newIngress() {
                         .endService()
                     .endDefaultBackend()
                     .addNewRule()
+                        .withHost(keycloak.getSpec().getHostnameSpec().getHostname())
```
Comment unrelated to this line. I believe we should not set this if `hostname == null`.

vmuzikar reviewed (changes requested):
Sorry, wrong button. :)

mabartos commented:
@vmuzikar Thanks for your review. I addressed your comments, and I'd like to try if the tests will pass in GHAs. I also need to make changes to the guides.

mabartos commented:
/rerun

mabartos commented:
/rerun

vmuzikar reviewed (changes requested):
@mabartos Thanks for the update!

Just a small heads up. There will be some conflicts with this: https://github.com/keycloak/keycloak/pull/14896

mabartos commented on operator/src/main/java/org/keycloak/operator/controllers/KeycloakDistConfigurator.java:
```diff
@@ -82,34 +83,32 @@ public void configureDistOptions() {
+            var specifiedPort = hostnameSpec.getPort();
 
-            envVars.addAll(enabledStrictHostname);
+            if (!specifiedPort.equals(Constants.KEYCLOAK_HTTP_PORT) || !specifiedPort.equals(Constants.KEYCLOAK_HTTPS_PORT)) {
+                status.addWarningMessage("It is not possible to change 'hostname-port' at this moment. " +
+                        "By default, we use Nginx Ingress, and the only possible ports are 80/443.");
+            }
```
Oh, you're right.

mabartos commented:
@vmuzikar Thanks! I'll update it, but it'd be good to wait until the HTTP opts are merged.

vmuzikar commented:
@mabartos #14896 was merged, I think we're now ready for a rebase.

pedroigor commented on operator/src/main/java/org/keycloak/operator/controllers/KeycloakDistConfigurator.java:
```diff
@@ -82,34 +83,32 @@ public void configureDistOptions() {
 
-            envVars.addAll(enabledStrictHostname);
+            if (!specifiedPort.equals(Constants.KEYCLOAK_HTTP_PORT) || !specifiedPort.equals(Constants.KEYCLOAK_HTTPS_PORT)) {
+                status.addWarningMessage("It is not possible to change 'hostname-port' at this moment. " +
```
I'm not sure if we can so strict about the `hostname-port` because you might have proxy in front of the cluster. I'm saying that based on past issues where there is a proxy chain (e.g.: https://github.com/keycloak/keycloak/issues/13261) and people exposing the server using a port other than the defaults.

For me we should either remove `hostname-port` from the CR or remove this validation.

mabartos replied:
Interesting issue with the proxy chain. Thanks for bringing this up. I'm in to remove `hostname-port` for now, as described below.

pedroigor commented on operator/src/main/java/org/keycloak/operator/controllers/KeycloakDistConfigurator.java:
```diff
@@ -82,34 +83,32 @@ public void configureDistOptions() {
-            envVars.addAll(enabledStrictHostname);
+            if (!specifiedPort.equals(Constants.KEYCLOAK_HTTP_PORT) || !specifiedPort.equals(Constants.KEYCLOAK_HTTPS_PORT)) {
+                status.addWarningMessage("It is not possible to change 'hostname-port' at this moment. " +
+                        "By default, we use Nginx Ingress, and the only possible ports are 80/443.");
```
We can't assume Nginx controller is the default.

pedroigor commented on operator/src/main/java/org/keycloak/operator/controllers/KeycloakIngress.java:
```diff
@@ -97,8 +97,8 @@ private Ingress newIngress() {
 
-        if (!keycloak.getSpec().isHostnameDisabled()) {
-            ingress.getSpec().getRules().get(0).setHost(keycloak.getSpec().getHostname());
+        if (keycloak.getSpec().getHostnameSpec() != null && keycloak.getSpec().getHostnameSpec().getHostname() != null) {
```
I also agree that setting the `hostname` is enough most of the time.

But we also have the `hostname-url` option to force a base URL to frontend URLs. That said, I'm not sure if you want to resolve the host from the `hostname-url` too in case this option is set.

mabartos replied:
Good point. 

I'd say we could parse the hostname from the `hostname-url`, if the `hostname` is not specified. Otherwise, prefer only the `hostname`. 

If both are specified, it's the responsibility of the user/dev to manage them, right?!

pedroigor reviewed (commented):
Out of curiosity, the ingress host is set based on the hostname options, right?

If not set, what is the ingress host?

vmuzikar commented:
@pedroigor Thank you for your input! This made me wondering. Should we then for now omit the hostname-port and hostname-path from the CRD? It'd still be possible to use them via the serverConfiguration free-form field.

mabartos commented:
@vmuzikar @pedroigor Thanks guys for your comments. 

> If not set, what is the ingress host?
AFAIK, only some external IP is used. But TBH, I'm not sure. I can verify it. Good point!

As is described in the doc for [Ingress rules](https://kubernetes.io/docs/concepts/services-networking/ingress/#ingress-rules):
_An optional host. In this example, no host is specified, so the rule applies to all inbound HTTP traffic through the IP address specified. If a host is provided (for example, foo.bar.com), the rules apply to that host._

I agree that we could omit the `hostname-port` ATM, as you've suggested, and we can discuss it later as it won't be a breaking change afterward.

mabartos committed:
Add hostname options to Keycloak CR

Closes #14395

Co-authored-by: Václav Muzikář <vmuzikar@redhat.com>

vmuzikar reviewed (approved):
@mabartos LGTM, thank you!

Let's wait for the tests.

vmuzikar reviewed (changes requested):
@mabartos Unfortunately, there seem to be some test errors related to this PR. Could you please check it?

mabartos commented:
/rerun

mabartos commented:
@vmuzikar Model test failures are not related to this PR.
